accepted/2021/platform-dependencies/platform-dependencies.md

```diff
@@ -0,0 +1,433 @@
+
+#### Platform End of Support
+
+When support for a platform is removed from the product, the platform dependency model of future releases needs to be updated remove this platform.
+
+### Selecting a Model File
+
+The first thing a maintainer needs to do in order to edit the dependency model is to know which model file to update.
+
+In cases where a new dependency is added or support is added for a new platform, the maintainer should select the dependency model associated with the upcoming release in the https://github.com/dotnet/core/tree/master/release-notes location. In all likelihood, there will not yet be a folder for the upcoming release so one can be added as the placeholder for the future release notes and define the dependency model file there.
+
+In all other cases, it is the past release dependency files that need to be updated. For example, if support for Debian 10 ends, all dependency model files for all past releases that contain the `debian.10` rid should be updated to have that platform removed.
+
+### CLI Tool
+
+Because there can be numerous dependency model files that need to be updated for certain scenarios -- and to reduce human error -- a CLI tool will be created in order to perform certain operations across multiple dependency model files. This will be intended for scenarios that would otherwise be repetitive busywork and can easily be automated. Examples of such operations include [dependency name changes](#dependency-name-change) and [platform removal](#platform-end-of-support).
+
+#### CLI Syntax
+
+##### Override a dependency name
+
+```console
+dependency override [--path <path>] <dependency-type> <source-rid> <source-dependency-name> <target-rid> <target-dependency-name>
+```
+
+Options:
+
+* `path`: Root path location where dependency model files will be searched and updated. Default is the current working directory.
+
+Arguments:
+
+* `dependency-type`: type of the dependency to override
+* `source-rid`: rid of the dependency to override
+* `source-dependency-name`: name of the dependency to override
+* `target-rid`: rid of the platform to contain the dependency override
+* `target-dependency-name`: new name of the dependency
+
+Example:
+
+Overrides the libicu package from the default of libicu57 for all Debian platforms to libicu63 for Debian 10:
+
+```console
+dotnet-deps dependency override LinuxPackage debian libicu57 debian.10 libicu63
+```
+
+##### Remove a platform
+
+```console
+platform remove [--path <path>] [--force] <rid>
+```
+
+Options:
+
+* `path`: Root path location where dependency model files will be searched and updated. Default is the current working directory.
+* `force`: Forces the platform to be removed even if it has child platforms.
+
+Arguments:
+
+* `rid`: rid of the platform to remove
+
+Example:
+
+Removes the debian.10 platform:
+
+```console
+dotnet-deps platform remove debian.10
+```
```

Comment:
I've looked at Node.js and Python so far and have not come across anything formalized like this.

@omajid - Have you or anyone else at RedHat come across any other platforms that do anything similar to this?

@tianon - In your role maintaining Docker Hub's official images, have there been any platforms out there that have created a standalone, machine-readable description of the platform's package dependencies such that it can be used to maintain the packages installed by the Dockerfile?  Or has it always just been the Dockerfile being the source of truth?  To give you a very brief summary of what is being proposed here for .NET, there would be a JSON file which formalizes .NET's platform dependencies, including Linux packages.  That JSON file could be used to do transformations into or validation of Dockerfiles, as an example.  The intention is to have a standalone, independent source of truth from which other assets could be derived.

Reply:
This is really ambitious -- I'm not aware of anyone trying to standardize this sort of metadata.

If I'm understanding correctly, you're trying to come up with a standard identifier for things like `libargon2-dev` in Debian/Ubuntu vs `argon2-dev` in Alpine etc?

I think a really big challenge you'll likely run into is that these things aren't always one-to-one mapped -- there are many cases where distributions provide multiple variants (see the `-dev` packages in https://packages.debian.org/source/sid/curl for example) or even split different binaries across packages differently.  In addition, these things tend to move between packages over time too (like how Debian's `btrfs-progs` used to also contain the development headers until the dedicated `libbtrfs-dev` package was introduced).  Even the "upstream project" often gets split differently in different distros (where some have a single source package and others will end up with multiple source packages representing the same thing).

A different approach to trying to identify/recognize all the explicit packages might be to identify commands and specific header/`pkg-config` files necessary, but even that's going to be a bit of guesswork (and if it isn't automated to some extent, will struggle with bitrot), and that list isn't always straightforward to come up with, even for a human (like how difficult it would be to determine in an automated way that a program ends up invoking `git` at runtime, or that it uses `dlopen` to load a shared library dynamically).

There's also going to be different classes of dependencies -- obvious ones are build vs runtime, but even at runtime there are some dependencies that are more "required" than others (see https://www.debian.org/doc/debian-policy/ch-relationships.html for an example of how Debian handles this with `Depends` vs `Recommends` vs `Suggests` etc).

Reply:
> Have you or anyone else at RedHat come across any other platforms that do anything similar to this?

I can try asking around, but I have not seen anything similar to this. Most upstreams are less disciplined than this, I think :smile:

Reply:
Thanks, @tianon

> If I'm understanding correctly, you're trying to come up with a standard identifier for things like `libargon2-dev` in Debian/Ubuntu vs `argon2-dev` in Alpine etc?

Not really a standard identifier.  There's no attempt to map 1:1 between distros or anything like that.  Package dependencies can be defined for each distro independently.  It's really no different than the dependencies that are described in [documentation form](https://docs.microsoft.com/dotnet/core/install/linux), it's just that they're defined in a machine-readable form using a common schema.

> There's also going to be different classes of dependencies -- obvious ones are build vs runtime

Yes, that's [accounted for](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R22-R23) here by having separate models between them.

>  but even at runtime there are some dependencies that are more "required" than others (see https://www.debian.org/doc/debian-policy/ch-relationships.html for an example of how Debian handles this with Depends vs Recommends vs Suggests etc).

This is sort of addressed by the [dependency usage](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R108-R111) concept. It's less prescriptive than the depends/recommends/suggests model, allowing for the dependency to be associated with specific dev or app scenarios.